GameLogic/python_emulator.py

```diff
@@ -0,0 +1,531 @@
+        self.dino_array = self.jump_array
+        self.dino_height, self.dino_width = self.dino_array.shape
+        self.dino_x, self.dino_y = 10, self.REF_AXIS - self.dino_height * self.GRID_SIZE
+        self.original_dino_y = self.dino_y
+        self.cactus_height, self.cactus_width = self.cactus_array.shape
+        self.cactus_y = self.REF_AXIS - self.cactus_height * self.GRID_SIZE
+        self.mid_bird_height, self.mid_bird_width = self.mid_bird_array.shape
+        self.mid_bird_y = self.MID_BIRD_AXIS - self.mid_bird_height * self.GRID_SIZE
+        self.curr_walk = self.UPDATE_WALK_FRAMES
+
+        # Dino Jump - Calculates time related values so it makes sense physically
+        self.time_to_peak = self.JUMP_TIME / 2
+        self.acceleration = (2 * self.JUMP_HEIGHT) / (self.time_to_peak**2)
+        self.init_velocity = self.time_to_peak * self.acceleration
+
+        # Obstacle Generation - Values for how fast generated obstacles move
+        self.COOLDOWN_TIME = self.JUMP_TIME + self.ADDITIONAL_COOLDOWN
+        self.min_moving_speed = self.dino_width / (self.JUMP_TIME * 5)
+        self.real_moving_speed = int(
+            self.min_moving_speed * self.ADDITIONAL_MOVING_SPEED
+        )
+
+        # Set up display
+        self.screen = pygame.display.set_mode((self.WIDTH, self.HEIGHT))
+        pygame.display.set_caption("DinoJump!")
+        self.clock = pygame.time.Clock()
+
+        # Game Attributes
+        self.start_game = False
+        self.is_right_foot_down = True
+        self.is_jumping = True
+        self.jump_start_time = 0
+        self.is_ducking = False
+        self.cannot_spawn_object = True
+        self.cooldown_start_time = 0
+        self.cactus_locs = []
+        self.mid_bird_locs = []
+        self.high_bird_locs = []
+
+    #############################################################################
+    # Drawing Functions
+    #############################################################################
+
+    # Draws the black line for the ground
+    def draw_ground(self):
+        for pixel in range(self.WIDTH):
+            pygame.draw.rect(
+                self.screen,
+                self.BLOCK_COLOR,
+                (pixel, self.GND_AXIS, self.GRID_SIZE, self.GRID_SIZE),
+            )
+
+    # Takes whichever sprite is dino_array currently and it's location and draws to the board
+    def draw_dino(self):
+        for row in range(self.dino_height):
+            for col in range(self.dino_width):
+                if self.dino_array[row, col] == 1:
+                    pygame.draw.rect(
+                        self.screen,
+                        self.BLOCK_COLOR,
+                        (
+                            self.dino_x + col * self.GRID_SIZE,
+                            self.dino_y + row * self.GRID_SIZE,
+                            self.GRID_SIZE,
+                            self.GRID_SIZE,
+                        ),
+                    )
```

Comment:
I noticed the pygame `draw.rect` method is called several times in each game loop with similar parameters. You can consider wrapping it into a function called something like `draw_pixel` that takes in x and y values, avoiding the need to repeat `self.screen`, `self.BLOCK_COLOR` and `self.GRID_SIZE` every time and improve readability.